### PR TITLE
[cross_file] Add XFileImage ImageProvider to show XFiles as images

### PR DESCRIPTION
**Pull request by ueman**
This adds an [ImageProvider](https://api.flutter.dev/flutter/painting/ImageProvider-class.html) which takes an XFile.
This enables people to use XFiles directly as an Image source.

I've added tests which are similar to the existing ImageProvider tests in the flutter/flutter repo. If that's not enough please let me know and I'll add missing test cases.

Fixes https://github.com/flutter/flutter/issues/92472

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/cross_file/test/image_data.dart [hunk 1]
```diff
@@ -0,0 +1,132 @@
+/// A 50x50 blue square png.
+const List<int> kBlueSquarePng = <int>[
```
Comment (ueman) [inline packages/cross_file/test/image_data.dart:2]: This is copied from https://github.com/flutter/flutter/blob/125545da322048565911bf81118cce60f05036ce/packages/flutter/test/image_data.dart#L6

File: packages/cross_file/test/x_file_image_test.dart [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import 'package:cross_file/cross_file.dart';
+import 'package:cross_file/src/x_file_image.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'image_data.dart';
+
+void main() {
+  test('XFileImage sets tag', () async {
```
Comment (ueman) [inline packages/cross_file/test/x_file_image_test.dart:13]: Similar to https://github.com/flutter/flutter/blob/125545da322048565911bf81118cce60f05036ce/packages/flutter/test/painting/image_provider_test.dart#L161-L168

File: packages/cross_file/test/x_file_image_test.dart [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    final XFile file = XFile('', bytes: bytes);
+    final XFileImage provider = XFileImage(file);
+
+    final MultiFrameImageStreamCompleter completer =
+        provider.load(provider, _decoder) as MultiFrameImageStreamCompleter;
+
+    expect(completer.debugLabel, 'XFileImage(${describeIdentity(file)})');
+  });
+
+  test('Resize image sets tag', () async {
```
Comment (ueman) [inline packages/cross_file/test/x_file_image_test.dart:24]: Similar to https://github.com/flutter/flutter/blob/125545da322048565911bf81118cce60f05036ce/packages/flutter/test/painting/image_provider_test.dart#L179-L188

**Comment by stuartmorgan** [comment]
@ditman Thoughts on whether we want to add this to the package? It could also be something that's left for a separate package to implement, which is where I'm leaning; a file abstraction and a Flutter image provider are very different layers of the software stack.

**Comment by ditman** [comment]
@stuartmorgan totally agree! this probably should be its own package, named `cross_file_image_provider` or whatever @ueman wants to call it!

Something that we could do in `cross_file` is to have a list in the README with these expansions that have been built on top of it to try and keep the duplication low? What do you think @ueman?

**Comment by ueman** [comment]
> What do you think @ueman?

What you're suggesting sounds reasonable. I'll change it accordingly.

@stuartmorgan @ditman Are the current tests enough?

**Comment by stuartmorgan** [comment]
> I'll change it accordingly.
> 
> @stuartmorgan @ditman Are the current tests enough?

When I suggested a separate package, I actually meant a third-party package. I don't think there's a compelling need for this adapter to be a Flutter-team-owned package. `cross_file` is here because it's a core part of the implementation of several of our plugins; no Flutter-team-owned code would use `XFileImage` as far as I know.

**Comment by ditman** [comment]
Note that I just found about this (from an unrelated issue):

* https://pub.dev/packages/cross_file_image

Maybe @ueman and @qiuxiang can collaborate? And we can link to `cross_file_image` from `cross_file` (so nobody else tries to do an ImageProvider with XFile? :))

**Comment by qiuxiang** [comment]
Yeah, `cross_file_image` is the `ImageProvider` for `XFile` that slightly modified from `FileImage`. Currently used in my projects and works fine. But lack of tests.

**Comment by ueman** [comment]
@qiuxiang It seems to me, that you should be able to just use the tests from this PR.

Thanks all. I'll close this then.

**Pull request closed without merging**